### ISSUE TITLE
Similarities for Predicting Drug-Target Interactions

### ISSUE THREAD

**Issue by saish20** [pz-white/COM6911-DTI-2021#3]
Samuel has created two similarity matrices, one each for drugs and target using this [paper](https://www.bic.kyoto-u.ac.jp/pathway/Files/kdd13.pdf) and Pykale model; I have attached the notebook. 
We, however, are unsure as to how we are supposed to leverage these matrices into our model to predict the interaction. Could you guide us on how we move from here on?
[Similarity Matrix.pdf](https://github.com/pz-white/COM6911-DTI-2021/files/6378086/Similarity.Matrix.pdf)

**Comment by SamuelMidgley** [comment]
![image](https://user-images.githubusercontent.com/80775550/116108904-3d6be900-a6ac-11eb-9a59-67c10fe01ba3.png)

For more clarification on this, For Matrix factorisation you update the weights of A and B to better approximate Y (Figure 1).  Then for a similarity based approach Figure 2 then suggests A and B are  also related to the similarity matrix for drug and target respectively. If this is the case then how do you go about updating the weights for A and B?  It explains it in Section 2.3 but I’m struggling to understand the math. Any help would be greatly appreciated. 
Paper referring to https://www.bic.kyoto-u.ac.jp/pathway/Files/kdd13.pdf

**Comment by pz-white** [comment]
Hi @saish20 @SamuelMidgley 
This paper used the ALS algorithm to optimize parameters and it required to calculate the first derivative for the matrix A, B and weight vectors w_d and w_j. In principle it is same as your previous MF optimization but this loss funtion is more complex. Here I have two advices:
- As the loss funtion has been illustrated in section 2.2, you can write the each paramter's derivative and finally you will find it is same as the format in section 2.3.
- If it is still hard to understand the derivative in section 2.3, I remommend you use pytorch/tensorflow and you can only need to focus on the forward propgation (i.e. how to calculate the loss function). Then you can use the gradient-based optimizer (e.g. Adam/SGD) to automatically optimize the parameters instead writing backpropgation algorithm by yourself. The method is more easy.
I'll explain it more on our group meeting if you're still confused about the loss function, thanks.

**Comment by SamuelMidgley** [comment]
Hi @pz-white,
Would it be possible for you to check the code I have uploaded to [Delta-MFwithCS](https://github.com/pz-white/COM6911-DTI-2021/blob/Delta-MFwithCS/Delta/MF%20with%20CS.ipynb), I tried implementing the method into pytorch and was hoping if you could confirm whether I had done it right?

Thank you

**Comment by saish20** [comment]
![image](https://user-images.githubusercontent.com/17896055/116994602-d44e2c00-acd0-11eb-90bd-8211d14a2c8a.png)

**Comment by pz-white** [comment]
Hi @saish20 @SamuelMidgley, about the cos_matrix function:
- I don't think the integer label encoding should be used for cos calculation as it is a non-normalized vector. What's the method of similarity matrix calculation in original paper? Why don't you follow it?
- Please ignore IC50 firstly and at least successfully run Ki/Kd data.
- I don't see the execution of MSCMF model in your notebook, have you achieved it?

**Comment by pz-white** [comment]
You should follow the similarity matrices in sec 3.2.2 of MSCMF paper, you can find that the CS and GS matrices are same as the methods in the paper I shared in checkbox before. And the two matrices are the most popular and standard similarities for drugs/proteins. I suggest that you can achieve them firstly and then feed into your MSCMF model. After finishing the whole workflow, then we can consider to add other matrices.

**Comment by SamuelMidgley** [comment]
Hi @pz-white,
Thank you for looking through the code, I used cosine similarity on the integer label encoding because I was unable to find the similarity matrix method that they used but now you have shared the paper I will try and implement them correctly. 

`> I don't see the execution of MSCMF model in your notebook, have you achieved it?`  - Is this not achieved in the forward pass of that notebook? The code is the highlighted section attached.
![image](https://user-images.githubusercontent.com/80775550/117002039-f9479c80-acda-11eb-814d-eb98695931e6.png)

Thanks again for looking through the code and we'll try and work through the points suggested.

**Comment by pz-white** [comment]
I see, thanks for explaining. I missunderstand it as the class is called MatrixFactorization.
